Bugfix constraint migration fix

objecttothis opened the pull request:
- Migrate Sessions table to prevent problems with migrating populated databases.
- Refactor a few function names for PSR-12 compliance.
- Fix bug in migration causing failure due to trying to drop indexes with associated foreign key constraints. This bug was causing 3.3.2 databases to fail migration. #4202 
- migration to change constraint back to ON DELETE RESTRICT
20250213000000_fix_attributes_cascading_delete.php reverted a change on the ospos_attribute_links_ibfk_1 and ospos_attribute_links_ibfk_2 constraints on the ospos_attribute_links table to fix a bug with not being able to delete a dropdown value. The problem is that these need to be ON DELETE RESTRICT in order to use generated_unique_column and attribute_links_uq3 to enforce no duplication of attribute links.  The solution to this (in this PR) is to change them back to ON DELETE RESTRICT and programmatically cascade delete before trying to remove an entry.  This way we can enforce uniqueness AND deleting dropdown rows.
- Add indices to ospos_inventory and ospos_attribute_links which were causing the items view queries to time out when attribute_links was large and the hardware wasn't fast.

closing #4237
closing #4251

jekkos commented on app/Database/Migrations/20240630000001_fix_keys_for_db_upgrade.php:
```diff
@@ -54,33 +54,91 @@ public function down(): void
-    private function create_primary_key(string $table, string $index): void
+    private function createPrimaryKey(string $table, string $index): void
     {
         $result = $this->db->query('SELECT 1 FROM information_schema.columns WHERE table_schema = DATABASE() AND table_name= \'' . $this->db->getPrefix() . "$table' AND column_key = '$index'");
```
this method looks very generic, could it be moved to the db helper?

objecttothis replied:
Possibly.  I'll look into it.

objecttothis replied:
I'm still going to extract this.  The rest of the code is ready for review.

objecttothis replied:
Done

jekkos commented on app/Database/Migrations/20240630000001_fix_keys_for_db_upgrade.php:
```diff
@@ -54,33 +54,91 @@ public function down(): void
-    private function index_exists(string $table, string $index): bool
+    private function deleteIndex(string $table, string $index): void
+    {
+        if ($this->indexExists($table, $index)) {
```
same with this one, I bet we can reuse those in other migrations

objecttothis replied:
Sounds good.  I'll look into it.

objecttothis replied:
I'll do this one also.  I believe we already have a migration helper where it can go.

objecttothis commented on app/Database/Migrations/20250213000000_fix_attributes_cascading_delete.php:
```diff
@@ -14,7 +14,12 @@ class Migration_Attributes_fix_cascading_delete extends Migration
     {
         helper('migration');
+
+        $this->db->query("ALTER TABLE `ospos_attribute_links` DROP INDEX `attribute_links_uq3`");
```
I hear you, but anyone with MySQL would be permanently unable to upgrade their version of OSPOS because this migration fails on those systems without this statement. I suppose the way to make sure that everyone ends up with the same database schema would be another migration which either drops it for everyone or adds it back in for people who migrated through this change.  I am working on a code change which will add the the index back in and revert the constraint changes made here to ON DELETE RESTRICT but without the bug of not being able to delete a dropdown selection item.  The short version of this is that ON DELETE RESTRICT is needed on the constraints in order to enforce uniqueness to attribute_links, so to still get the effects of ON DELETE CASCADE we have to programmatically implement ON DELETE CASCADE.

objecttothis commented on app/Controllers/Attributes.php:
```diff
@@ -230,18 +230,6 @@ public function getView(int $definition_id = NO_DEFINITION_ID): void
         echo view('attributes/form', $data);
     }
 
-    /**
```
This method wasn't actually called anywhere.

jekkos replied:
Beware it might be in use but called from the frontend. I might have cone across it when I fixed the bug where we were not able to delete an attribute droodown value.

objecttothis replied:
PHPStorm also doesn't think it's called anywhere but it wouldn't find AJAX calls anyway.  String Searches don't find it.

jekkos replied:
Ok but let's check if the dropdown delete still works then after

objecttothis replied:
I checked it in my testing and it still worked to delete a drop-down choice. I confirmed also that it did not delete attribute link rows attached to sales or receivings. The cascading delete of items with that drop-down choice also works.

I don't think I tested deleting an entire drop-down attribute so we need to test that.

objecttothis commented on app/Database/Migrations/20210422000001_remove_duplicate_links.php:
```diff
@@ -35,6 +35,7 @@ private function migrate_duplicate_attribute_links(): void
         $builder->select('item_id, definition_id, attribute_id, COUNT(*) as count');
         $builder->where('sale_id', null);
         $builder->where('receiving_id', null);
+        $builder->where('item_id IS NOT NULL');
```
This is a retroactive fix because existing OSPOS MySQL databases can't get past it otherwise.

jekkos replied:
Thanks for that

objecttothis commented on app/Database/Migrations/sqlscripts/3.4.1_attribute_links_unique_constraint.sql:
```diff
@@ -0,0 +1,16 @@
+ALTER TABLE `ospos_attribute_links` DROP CONSTRAINT `ospos_attribute_links_ibfk_1`;
```
@jekkos regarding your objection above about retroactively changing the migration to drop attribute_links_uq3 and how MariaDB users will have already run that, the way to fix it is by this migration.
There are two scenarios: MySQL DBMS and MariaDB DBMS.  Those on the MySQL DBMS will not have generated_unique_column when it hits line 6 in this migration and people with MariaDB DBMS will have it.  I haven't written it yet, but I need to write the php up() function that calls this sql to instead programmatically drop the generated_unique_column if it exists so that we don't get an error on MariaDB databases that the column exists.

jekkos replied:
That sounds good

BudsieBuds commented on .editorconfig:
These changes should probably be removed, superseded by #4201.

jekkos commented on app/Views/attributes/form.php:
```diff
@@ -181,7 +181,7 @@
         else
         {
-            $.post('<?= esc("$controller_name/delete_attribute_value/") ?>', {definition_id: definition_id, attribute_value: value});
+            $.post('<?= esc("$controller_name/DeleteDropdownAttributeValue/") ?>', {definition_id: definition_id, attribute_value: value});
```
Does it work with the Delete first char uppercase?

objecttothis replied:
I believe so.  I'll test it.  I was following the capitalization of the function name which is I believe is PostDeleteDropdownAttributeValue().  It's possible the router may not care.  In this case should we just make it all lowercase?

jekkos replied:
I though it was camelcase from the view but I could be wrong

objecttothis replied:
PascalCase from the view is working as it is.  It might be case insensitive, but IMO PascalCase matches the actual function name most precisely.

objecttothis commented on app/Models/Attribute.php:
```diff
@@ -41,6 +41,38 @@ class Attribute extends Model
     public const SHOW_IN_SALES = 2;
     public const SHOW_IN_RECEIVINGS = 4;
 
+    // ==============
```
@jekkos I would like to move toward organizing the models with these section headings.  Right now the models are pretty disorganized.  IDEs like PHPStorm can help but I'd still like to move things around.  C# has `#region foo` but PHP doesn't have anything like that.  What do you think?

objecttothis replied:
I see your point.  PHPStorm does have a 'structure' window which shows the function signatures and this is very helpful, but as you can see from the sale model below that the list appears in the order that the function appears in the model, which is a hot mess.  This means that if I'm looking to see if we have a specific function, I often have to scan the entire list to find it.  I don't feel extremely strongly about using comments to separate sections, but it really would be useful if our classes all followed a similar ordering. Something like:
Fields
Constructors
Create
Read
Update
Delete
Private
Static
![image](https://github.com/user-attachments/assets/622136a3-590d-47f6-8b86-cd78cf53f7f7)

Thoughts?

jekkos replied:
Agree I would try to stick to an order in which we define method type or fields then instead of using comments. It's more consistent in our case and indeed, we have a lot of methods in a controller class. So better to deal with it this way

objecttothis commented on app/Models/Attribute.php:
```diff
@@ -987,4 +1025,38 @@ public function get_attributes_by_definition(int $definition_id): array
         return $builder->get()->getResultArray();
     }
+
+// ==============
```
This heading would be for private Model functions.

jekkos replied:
I'd try to code to structure things logically.. having to add comments might be a code smell (according to clean code guidelines). It might just indicate naming is not good or classes are not sticking to single resonsibility principle.

objecttothis commented:
@jekkos, I think the commit history is a little rough after rebasing on the master.  When I merge, per usual I'm planning to do a squash and merge with a commit message of this to clean everything up:

```
- Refactored function names for PSR-12 compliance
- Programmatically cascade delete attribute_link rows when a dropdown attribute is deleted but leave attribute_link rows associated with transactions.
- Added `WHERE item_id IS NOT NULL` to migration to prevent failure on MySQL databases during migration
- Retroactive correction of migration to prevent MySQL databases from failing.
- Refactored generic functions to helper
- Reverted attribute_links foreign key to ON DELETE RESTRICT which is required for a unique constraint on this table. Cascading deletes are now handled programmatically.
- Migration Session table to match Code Igniter 4.6
- Add index to attribute_links to prevent query timeout in items view on large databases
- Added overridePrefix() function to the migration_helper. Any time QueryBuilder is adding a prefix to the query when we don't want it to, this query can be used to override the prefix then set it back after you're done.
- Added dropAllForeignKeyConstraints() helper function.
- Added deleteIndex() helper function.
- Added indexExists() helper function.
- Added primaryKeyExists() helper function.
- Added recreateForeignKeyConstraints() helper function.
- Added CRUD section headings to the Attribute model.
- Replaced `==` with `===` to prevent type juggling.
- Removed unused delete_value function.
- Reworked deleteDefinition() and deleteDefinitionList() functions to delete rows from the attribute_links table which are associated.
- Added deleteAttributeLinksByDefinitionId() function
```

This PR is now ready for review.

objecttothis commented:
@jekkos the PHPLinter is failing on PHP 8.4 deprecations (https://github.com/opensourcepos/opensourcepos/actions/runs/15184079188/job/42700305258).  I do think we need to make these changes to remove deprecated code, but that needs to be a separate PR from this one.

jekkos commented:
We can keep the full history in the squashed commit message body. So typically a good commit message has a short first line describing what is is about, and then more details can be added after the first newline. 

I find it very useful to add info to the history this way, it's a bit of an ongoing documentation with every change that is done.

jekkos commented on app/Models/Attribute.php:
```diff
@@ -842,29 +874,16 @@ public function saveAttributeValue(string $attribute_value, int $definition_id,
-     * @param int $definition_id
-     * @return bool
-     */
-    public function delete_value(string $attribute_value, int $definition_id): bool
```
@objecttothis I double checked the code and found that this method is in use in the controller `Attributes.php:83`

jekkos replied:
Ah I see you replaced it with a another one, probably fine

jekkos commented on app/Views/attributes/form.php:
```diff
@@ -160,7 +160,7 @@
                 values.splice($.inArray(value, values), 1);
             } else {
-                $.post('<?= esc("$controller_name/delete_attribute_value/") ?>', {
+                $.post('<?= esc("$controller_name/DeleteDropdownAttributeValue/") ?>', {
```
It's this one that calls the method you removed here

jekkos commented on app/Database/Migrations/20250522000000_AttributeLinksUniqueConstraint.php:
```diff
@@ -0,0 +1,28 @@
+
+        execute_script(APPPATH . 'Database/Migrations/sqlscripts/3.4.1_attribute_links_unique_constraint.sql');
+
+        error_log('Migrating Attribute Links Unique Constraint');
```
Maybe one message is suffiicient?

jekkos reviewed (dismissed):
I'm not going to block the PR here at this point, I know these things take time to test. But I do have some minor remarks on the usage of comments and eventual removal of a duplicated log line. Beside that I think it looks good.

jekkos commented on app/Database/Migrations/20250522000000_AttributeLinksUniqueConstraint.php:
```diff
@@ -0,0 +1,28 @@
+    public function up(): void
+    {
+        error_log('Migrating Attribute Links Unique Constraint');
+
```
Great thanks for resolving this!

objecttothis commented:
@jekkos in the morning I'll remove the comments and extra log lines. I think I was just mimicking what was in other migrations.

I'm OK with just ordering class functions logically, but I think we should agree on an order and move functions to those locations as we work on the code. Doing it all here would make the commit pretty difficult to parse.

I'll squash and merge after pushing those commits.

objecttothis commented:
@jekkos I was taught a class order of

Fields
Constructors 
Create
Read
Update
Delete
Private methods

Does that work or a variation of it? If not feel free to propose a different order. I just get tired of wasting time not knowing exactly where a function of a certain time would generally be in a model.
